### DISCUSSION TITLE
Analytics and Reporting

### DISCUSSION THREAD

**Discussion by fforootd** [zitadel/zitadel#2162] [Ideas]
This discussion is to define the scope for two upcoming features (Analytics and Reporting).

Please check the threads below for additional inputs.

CC @mffap @livio-a @fgerschwiler @adlerhurst

**Comment by fforootd** [comment]
# Analytics (Threat Analytics)

This thread outlines the use cases, functions and features we think is relevant for analytics

## Use Cases

- Decrease the risk of account breaches
- Decrease the risk of account changes (data changes, access escalation)

## Features

- Check the users authN and re-authN against a model
- Check the users changes against a model
- Evaluate anonymous sessions against a model (e. g when the user is not logged in)

## Functions

- Feed Eventstream into model
- Create a query model with the underlying analytics model
- Provide this not only as function but instead also as API
- feedback loop to integrate the users and managers opinion if something is right or wrong

**Reply by mffap** [reply]
There should be an option to provide feedback on detected threats. Users or managers should be able to provide feedback on detected threats (eg, "Yes this was me" / "No this was not me"). This information should be stored to improve the model.

**Reply by fforootd** [reply]
Yeah a feedback loop makes total sense

**Reply by mffap** [reply]
@fgerschwiler @fforootd @adlerhurst getting a feedback on relevant audit events* to identify threats (true positives) is a necessary requirement for further models. I tried not to use terminology, and certain examples might be not feasible, but hopefully I can convey the intend here.
[*] here in a business sense, but could match with the technical implementation. For me this is similar to what is currently shown on the sidebar in console.

I propose the following in order of subjective importance and illustrations

**1/ Users should be able to flag and report suspicious events**

Given I have events in my audit log, when I can 'flag' these events as suspicious, then these events are market in the system-wide audit log and an email will be send to manager roles.

Examples

- Passwordless Login check succeeded
- Multifactor OTP check succeeded

**2/ Inform users/managers of sensitive actions**

Given there is a set of 'sensitive' events, when this action/event is triggered, then the relevant user and/or manager is informed and an audit event is created.

Examples for such events could be changes to... 
- MFA or authenticators
- password
- verified email (info to previous email)
- manager roles
- redirect urls for applications
- ...

**3/ Provide a system-wide audit report**

- Provide all audit events in a system-wide report
- Mark events flagged as suspicious
- Mark 'sensitive events'
- Allow users to filter, sort, and search the audit log

**4/ Create rule-based audit events**

When a audit event is created based on rules, then it should be visible in the audit report.

Examples for such audit events could be
- x sequential login (successful and unsuccessful) attempts during a period of time
- changes in user-agent or ip between login attempts during a period of time
- changes in user-agent or ip compared to active session
- (OpSec) rate-limited service accounts
- (OpSec) rate-limited login attempts
- users with significant more authorizations within an organization

**5/ Certain rules trigger tighter auth enforcement**

Given a set of 'risky events', when an audit event is created based on rules and this audit event is a 'risky event', then certain actions will be performed to mitigate the risk of a potential threat, and the 'risky event' is logged in the audit log (for feedback).

Mitigations might include
- require re-authentication
- deactivate user account

**Comment by fforootd** [comment]
# Reporting (Metrics and Audit)

This thread outlines the use cases, functions and features we think is relevant for reporting

## Use Cases

- As a customer I want to build dashboards and reports to see how many (and what)
  - users are active
  - users a app has
  - request my org handles
  - compliance and login violations happen
  - threats detected and mitigated
-  As a customer I want to be able to query the audit trail easily
  - If I detect a breach I want to time travel to the timeframe of the breach an check how things looked there

## Features

- Metrics API that provides easy access to build dashboards and reports (something like a promQL could be interesting)
- Audit API which allows for easy filtering of multiple resources at a given time
- We should also provide some metrics and more audit data into `console`

## Functions

- Stream events into a Metrics Model
- Stream events it a Audit Model (It is already possible to check all the events from a single resource)

**Reply by mffap** [reply]
Time dimensions
 - last 7 days vs. previous 7 days
 - last 24h vs. previous 24h
 
Metrics
- How many users registered to my organization
- How many successful logins
- Hom many failed logins (and in terms of successful vs. failed)
